src/serializer/Emitter.js

```diff
@@ -175,7 +198,29 @@ export class Emitter {
       this._residualFunctions.addFunctionUsage(val, this.getBodyReference());
       return undefined;
     } else if (val instanceof AbstractValue) {
-      if (val.hasIdentifier() && !this._declaredAbstractValues.has(val)) return val;
+      if (val.hasIdentifier()) {
+        const valSerializeBodyStack = this._declaredAbstractValues.get(val);
+        if (!valSerializeBodyStack) {
+          // Hasn't been serialized yet.
+          return val;
+        } else {
+          // The dependency has already been serialized(declared). But we may still have to wait for
+          // current generator body to be available, under following conditions:
+          // 1. Not delay initialization.
+          // 2. Not emitting in current active generator.(otherwise no need to wait)
+          // 3. Dependency's active ancestor generator body is lower in generator stack than current body.
+          const valActiveAncestorBody = this._getFirstAncestorGeneratorWithActiveBody(valSerializeBodyStack);
+          invariant(this._activeBodies.includes(valActiveAncestorBody));
+          if (!this._activeBodies.includes(this._body)) {
+            invariant(this._delayInitializations);
```

Comment:
And add a TODO to not just check whether the `this._delayInitializations` option is on, but pass through more information to check that `this._body` is a delayed initialization body.